### PR TITLE
support grpc plugin

### PR DESCRIPTION
**Pull request by ascrutae**

### CODE REVIEW

**Comment by ascrutae** [comment]
Here is screen snapshot of the stream call trace
wx20171016-142434 2x

Here is screen snapshot of stream call span info
wx20171016-234344 2x

Here is screen snapshot of  block call trace 
wx20171016-142617 2x

File: apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/vo/CachedObjects.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * Project repository: https://github.com/OpenSkywalking/skywalking
+ */
+
+package org.skywalking.apm.plugin.grpc.v1.vo;
+
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import org.skywalking.apm.agent.core.context.ContextSnapshot;
+
+public class CachedObjects {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/vo/CachedObjects.java]: Why named `CachedObjects`? Any object with status can be named this.

File: apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/vo/ServiceDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    private MethodDescriptor.MethodType methodType;
+    private String serviceName;
+
+    public ServiceDescriptor(MethodDescriptor descriptor) {
+        this.methodType = descriptor.getType();
+        String fullMethodName = descriptor.getFullMethodName();
+        this.serviceName = dealWithServiceName(fullMethodName) + "." + dealWithMethodName(fullMethodName);
+    }
+
+    private String dealWithServiceName(String requestMethodName) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/vo/ServiceDescriptor.java]: `dealWith` is too abstract, it is more likely a format.

File: apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/vo/ServiceDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        int splitIndex = requestMethodName.lastIndexOf("/");
+        return requestMethodName.substring(0, splitIndex);
+    }
+
+    private String dealWithMethodName(String requestMethodName) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/vo/ServiceDescriptor.java]: `dealWith` is too abstract, it is more likely a format.

File: apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/StreamClientOnCloseInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        MethodInterceptResult result) throws Throwable {
+
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        Object ret) throws Throwable {
+        AbstractSpan abstractSpan = ContextManager.activeSpan();
+        abstractSpan.tag("onNext.count", String.valueOf(((CachedObjects)objInst.getSkyWalkingDynamicField()).getOnNextCount()));
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/StreamClientOnCloseInterceptor.java]: Why no use constant/staticField?

File: apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/ServerCallOnReadyInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        CarrierItem next = contextCarrier.items();
+        while (next.hasNext()) {
+            next = next.next();
+            String contextValue = headerMap.get(next.getHeadKey());
+            if (!StringUtil.isEmpty(contextValue)) {
+                next.setHeadValue(contextValue);
+            }
+        }
+
+        final AbstractSpan span = ContextManager.createEntrySpan(cachedObjects.getRequestMethodName() + (cachedObjects.getMethodType() != MethodDescriptor.MethodType.UNARY ? "/StreamCall" : "/UnaryCall"), contextCarrier);
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/ServerCallOnReadyInterceptor.java]: Why no use constant/staticField?

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/ServerCallOnReadyInterceptor.java]: Is `UnaryCall` a gRPC inside name? I think it is from here: `MethodDescriptor.MethodType.UNARY`, right? But why do you choose this name? It is hard to understand. for normal user.

File: apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/ServerCallOnMessageInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import org.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.skywalking.apm.plugin.grpc.v1.vo.CachedObjects;
+
+public class ServerCallOnMessageInterceptor implements InstanceMethodsAroundInterceptor {
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        MethodInterceptResult result) throws Throwable {
+        CachedObjects cachedObjects = (CachedObjects)objInst.getSkyWalkingDynamicField();
+        ContextManager.createLocalSpan(cachedObjects.getRequestMethodName() + "/ResponseStreamObserver/OnNext");
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/ServerCallOnMessageInterceptor.java]: Why no use constant/staticField?

File: apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/ServerCallOnCloseInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        MethodInterceptResult result) throws Throwable {
+
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        Object ret) throws Throwable {
+        AbstractSpan abstractSpan = ContextManager.activeSpan();
+        abstractSpan.tag("onNext.count", String.valueOf(((CachedObjects)objInst.getSkyWalkingDynamicField()).getOnNextCount()));
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/skywalking/apm/plugin/grpc/v1/ServerCallOnCloseInterceptor.java]: Why no use constant/staticField?

**Pull request merged**